lib/charms/mongodb_libs/v0/machine_helpers.py

```diff
@@ -146,29 +154,64 @@ def generate_service_args(auth: bool, machine_ip: str, replset: str) -> str:
         "--bind_ip",
         f"localhost,{machine_ip}",
         # part of replicaset
-        "--replSet",
-        f"{replset}",
+        f"--replSet={config.replset}",
     ]
 
     if auth:
-        mongod_start_args.extend(
-            [
-                "--auth",
-                # keyFile used for authentication replica set peers, cluster auth, implies user
-                # authentication hence we cannot have cluster authentication without user
-                # authentication see:
-                # https://www.mongodb.com/docs/manual/reference/configuration-options/#mongodb-setting-security.keyFile
-                # TODO: replace with x509
-                "--clusterAuthMode=keyFile",
-                f"--keyFile={KEY_FILE}",
-            ]
-        )
+        mongod_start_args.extend(["--auth"])
+
+        if config.tls_external:
+            mongod_start_args.extend(
+                [
+                    f"--tlsCAFile={TLS_EXT_CA_FILE}",
+                    f"--tlsCertificateKeyFile={TLS_EXT_PEM_FILE}",
+                    # allow non-TLS connections
+                    "--tlsMode=preferTLS",
+                ]
+            )
+
+        # internal TLS can be enabled only if external is enabled
+        if config.tls_internal and config.tls_external:
+            mongod_start_args.extend(
+                [
+                    "--clusterAuthMode=x509",
+                    "--tlsAllowInvalidCertificates",
+                    f"--tlsClusterCAFile={TLS_INT_CA_FILE}",
+                    f"--tlsClusterFile={TLS_INT_PEM_FILE}",
+                ]
+            )
+        else:
+            # keyFile used for authentication replica set peers if no internal tls configured.
+            mongod_start_args.extend(
+                [
+                    "--clusterAuthMode=keyFile",
+                    f"--keyFile={KEY_FILE}",
+                ]
+            )
 
     mongod_start_args.append("\n")
     mongod_start_args = " ".join(mongod_start_args)
 
     return mongod_start_args
 
 
+def push_file_to_unit(parent_dir, file_name, file_contents) -> None:
+    """K8s charms can push files to their containers easily, this is the vm charm workaround."""
+    Path(parent_dir).mkdir(parents=True, exist_ok=True)
+    with open(file_name, "w") as write_file:
+        write_file.write(file_contents)
+
+    os.chmod(file_name, 0o400)
+    mongodb_user = pwd.getpwnam(MONGO_USER)
+    os.chown(file_name, mongodb_user.pw_uid, mongodb_user.pw_gid)
```

Comment:
tip: you can use [`shutil.chown`](https://docs.python.org/3/library/shutil.html#shutil.chown) to use an one liner